### PR TITLE
Add "forbidden traits" to professions 

### PR DESCRIPTION
**Pull request by np-vortex**
#### Summary
```SUMMARY: Features "Enable defining forbidden traits in profession .json configuration"```

#### Purpose of change
To allow for more complex combination of scenario and professions configuration in regards to locked and forbidden traits. 

#### Describe the solution
Adds handling of "forbidden_traits" in professions json config. 
Adds logic to filter out professions whose traits conflict with traits for the scenario. 
Adds a debug message on scenario select if scenario has an explicitly configured profession but there is a conflict in trait configuration between the scenario and a profession.
Debug message would appear if:

- scenario has a locked trait that is also a forbidden trait for a profession for that scenario
- scenario has a forbidden trait that is also a locked trait for a profession for that scenario
- scenario has a locked trait that conflicts with a locked trait for a profession for that scenario

> I would like to add checks for traits that override or cancel each other, basically what is done in player::has_conflicting_trait, but that would mean I should take that logic outside of src/player.cpp and make it reusable. 

I have added the extra checks. Should I attempt to refactor player::has_conflicting_trait to use more generic code to check for conflicts in traits?

#### Describe alternatives you've considered
Do not throw a debug message in game.
Do not check for conflicting traits at the moment of scenario selection or anywhere else in game, but do it outside of game by parsing json files and detecting conflicting configuration, maybe even during running of tests.
Instead of removing a profession from menu in game, make it greyed out and display a message if it is clicked on with an explanation of why the player can not choose this profession. This is more complex.

#### Testing
Chose a scenario Y that has profession profession Z enabled. Added a trait X as a locked trait to a scenario Y, added trait X as a forbidden trait to a profession Z, started a new game, picked scenario Y, observed debug message about conflict in trait configuration between scenario Y and profession Z.

#### Additional context
Even before this change it was possible that scenario and  profession trait configuration was conflicting, as a scenario could have had a forbidden trait that was a locked trait for a profession for that scenario. I assume this was avoided by paying attention to not create such conflicts in the first place. Because I see no checks in the code that look for such conflicts, I am not sure If I should even attempt to create them. 

This feature was probably suggested before, but I first observed it in https://github.com/CleverRaven/Cataclysm-DDA/pull/34808#issuecomment-544310508

### CODE REVIEW

File: src/scenario.cpp [hunk 1]
```diff
@@ -232,16 +232,24 @@ std::vector<string_id<profession>> scenario::permitted_professions() const
     for( const profession &p : all ) {
         const bool present = std::find( professions.begin(), professions.end(),
                                         p.ident() ) != professions.end();
+
+        bool conflicting_traits = this->scenario_traits_conflict_with_profession_traits(p);
```
Comment (BevapDin) [inline src/scenario.cpp]:
```suggestion
        bool conflicting_traits = scenario_traits_conflict_with_profession_traits(p);
```

File: src/mutation.cpp [hunk 2]
```diff
@@ -1404,3 +1404,34 @@ void test_crossing_threshold( Character &guy, const mutation_category_trait &m_c
...
+
+bool are_same_type_traits( const trait_id &trait_a, const trait_id &trait_b )
+{
+    return contains_trait( get_mutations_in_types( trait_a->types ), trait_b);
+}
+
+bool contains_trait( std::vector<string_id<mutation_branch>> traits, const trait_id &trait )
+{
+    return std::find(traits.begin(), traits.end(), trait) != traits.end();
+}
```
Comment (anothersimulacrum) [inline src/mutation.cpp]:
```suggestion
}

```

**Comment by kevingranade** [comment]
Needs astyling.

**Comment by ZhilkinSerg** [comment]
Needs some json example - maybe add some trait restrictions to a challenge?

**Pull request merged**